### PR TITLE
don't move expensive function calls into loops

### PR DESCRIPTION
**Pull request by jsteemann**
### Scope & Purpose

Don't move potentially expensive AQL function calls into loops in the `remove-unnecessary-calculations-rule`.

For example, in the query
``` 
LET x = NOOPT(1..100)
LET values = SORTED(x)
FOR j IN 1..100 
  FILTER j IN values
  RETURN j
```
there is only one use of the `values` variable. So the optimizer can remove that variable and replace the filter condition with `FILTER j IN SORTED(x)`. However, that would move the potentially expensive function call into the inner loop, which could be a pessimization.

With this PR, the optimizer will not move the calculation of `values` into the loop when it merges calculations in the `remove-unnecessary-calculations` optimizer rule.

The original motivation for this change was this query:
``` 
LET values = (some subquery whose results are only known at runtime)
FOR doc IN collection
  FILTER doc._key IN values
  RETURN doc
```
Here, the optimizer decided in the `sort-in-values` rule that the right-hand operand of the `IN` operation is potentially huge, so it turned the linear search `IN` into a binary search `IN`. This requires sorted input, so the optimizer transformed the plan to
``` 
LET sub = (some subquery whose results are only known at runtime)
LET values = SORTED_UNIQUE(sub)
FOR doc IN collection
  FILTER doc._key IN values
  RETURN doc
``` 
This is still fine. However, then the `remove-unnecessary-calculations` rule came along and transformed this into
``` 
LET sub = (some subquery whose results are only known at runtime)
FOR doc IN collection
  FILTER doc._key IN SORTED_UNIQUE(sub)
  RETURN doc
``` 
which moved the sorting into the loop. This is definitely not optimal.
The change in this PR will prevent that from happening.

- [x] Bug-Fix for *devel-branch* (i.e. no need for backports?)
- [x] The behavior in this PR can be (and was) *manually tested* (support / qa / customers can test it)
- [x] The behavior change can be verified via automatic tests

### Testing & Verification

This PR adds tests that were used to verify all changes:

- [x] Added new **integration tests** (i.e. in shell_server / shell_server_aql)

http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10180/

### CODE REVIEW

File: arangod/Aql/OptimizerRules.cpp [hunk 1]
```diff
@@ -2867,14 +2882,29 @@ void arangodb::aql::removeUnnecessaryCalculationsRule(Optimizer* opt,
...
         }
-
-        if (!n->isInInnerLoop() && rootNode->callsFunction() && other->isInInnerLoop()) {
-          // original expression calls a function and is not contained in a loop
-          // we're about to move this expression into a loop, but we don't want
-          // to move (expensive) function calls into loops
-          continue;
+        
+        if (rootNode->callsFunction() && other->isInInnerLoop()) {
+          if (n->getLoop() != other->getLoop()) {
```
Comment (mchacki) [inline arangod/Aql/OptimizerRules.cpp:2887]: (This is micro optimization)
I am actually not convinced that this matters as soon as you read one Document, just for the sake of completeness.

getLoop() is implemented using a "walker" over dependencies, this potentially has some computation time involved.
Here we need n->getLoop() twice. and looking at isInInnerLoop() impleentation of other we also need the loop of other twice.
So we could save some lookups if we fetch both before their ifs.

I am also not sure if compiler is clever enough to figure out that n->getLoop() is constant and only required once here, making my comment absolutely obsolete.

Comment (jsteemann) [reply arangod/Aql/OptimizerRules.cpp:2887]: I think the compiler will evaluate this multiple times, so I can fix it.

**Comment by jsteemann** [comment]
Restarted Jenkins after review changes:
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10182/

**Comment by jsteemann** [comment]
http://172.16.10.101:8080/view/PR/job/arangodb-matrix-pr/10191/

**Pull request merged**